### PR TITLE
mvps: improve object move

### LINKED ISSUE #361: Piston with node doesn't push players/entities with most drawtypes
If you attach a node with a non-default drawtype to the end of a piston, it won't push entities, as shown below:

screenshot_20170907_025221
screenshot_20170907_025232

If there is no node or a full block node, the player/entity will be pushed as expected.

With a nodebox of any size or shape (or any other non-default drawtype e.g. raillike, plantlike), the node will simply clip through any players/entities standing on them, failing to push them up and potentially trapping them.

Edit: It seems that all nodes are treated as if they are full nodes, so if there is a 1/16th slab on a piston, it won't move the entity unless that entity is 2 nodes away from the piston, even if there is space in the adjacent node.

### PR DESCRIPTION
**Pull request by DS-Minetest**
Objects with big collision boxes are moved.
Objects in stack are moved, not only at front. Ergo fixes #361.
~~Objects lying on top of the nodestack aren't moved anymore. Bad?~~ Edit: fixed.

### CODE REVIEW

File: mesecons_mvps/init.lua [hunk 1]
```diff
@@ -205,37 +205,50 @@ end)
...
-				obj:setpos(np)
+		if ok then
+			local ent = obj:get_luaentity()
+			if not ent or not mesecon.is_mvps_unmov(ent.name) then
+				local np = vector.add(obj_pos, dir)
+				-- Move only if destination is not solid or object is inside stack:
+				local nn = minetest.get_node(np)
+				local node_def = minetest.registered_nodes[nn.name]
+				if (node_def and not node_def.walkable) or
+						math.abs(obj_pos[dir_k]) <= #nodestack - 0.5 + math.abs(pos[dir_k]) then
```
Comment (numberZero) [inline mesecons_mvps/init.lua]: What if signs of `pos[dir_k]` and `obj_pos[dir_k]` are different?

File: mesecons_mvps/init.lua [hunk 2]
```diff
@@ -205,37 +205,50 @@ end)
...
+		local obj_pos = obj:get_pos()
+		local cbox = obj:get_properties().collisionbox
+		local min_pos = vector.add(obj_pos, vector.new(cbox[1], cbox[2], cbox[3]))
+		local max_pos = vector.add(obj_pos, vector.new(cbox[4], cbox[5], cbox[6]))
+		local ok = true
+		for k, v in pairs(pos) do
+			local edge1, edge2
+			if k ~= dir_k then
+				edge1 = v - 0.5
+				edge2 = v + 0.5
```
Comment (numberZero) [inline mesecons_mvps/init.lua]: > Objects lying on top of the nodestack aren't moved anymore. Bad?

You mean, when pushing horizontally? That’s acceptable, but may be fixed by increasing (slightly) constant here, if I understand the code correctly.

File: mesecons_mvps/init.lua [hunk 3]
```diff
@@ -205,37 +205,50 @@ end)
...
+			else
+				edge1 = v - 0.5 * dir_l
+				edge2 = v + (#nodestack + 0.5) * dir_l
+				-- Make sure, edge1 is more than edge2:
+				if edge1 > edge2 then
+					edge1, edge2 = edge2, edge1
+				end
+			end
```
Comment (numberZero) [inline mesecons_mvps/init.lua]: `edge`s may be pre-calculated

File: mesecons_mvps/init.lua [hunk 4]
```diff
@@ -205,37 +205,50 @@ end)
...
-
-	-- Move objects lying/standing on the stack (before it was pushed - oldstack)
-	if tonumber(minetest.setting_get("movement_gravity")) > 0 and dir.y == 0 then
-		-- If gravity positive and dir horizontal, push players standing on the stack
-		for _, n in ipairs(nodestack) do
-			local p_above = vector.add(n.pos, {x=0, y=1, z=0})
-			local objects = minetest.get_objects_inside_radius(p_above, 1)
-			for _, obj in ipairs(objects) do
-				table.insert(objects_to_move, obj)
+	for id, obj in pairs(minetest.object_refs) do
```
Comment (numberZero) [inline mesecons_mvps/init.lua]: Seems horribly, but that’s the only way; MT itself does the same, all the time...

File: mesecons_mvps/init.lua [hunk 5]
```diff
@@ -205,37 +205,50 @@ end)
...
-			local np = vector.add(obj:getpos(), dir)
-
-			--move only if destination is not solid
-			local nn = minetest.get_node(np)
-			if not ((not minetest.registered_nodes[nn.name])
-			or minetest.registered_nodes[nn.name].walkable) then
-				obj:setpos(np)
+		if ok then
+			local ent = obj:get_luaentity()
+			if not ent or not mesecon.is_mvps_unmov(ent.name) then
```
[the remark this answers is not in the document]
Comment (DS-Minetest) [reply mesecons_mvps/init.lua]: `not (ent and …)`

File: mesecons_mvps/init.lua (hunk 3 above)
Comment (DS-Minetest) [reply mesecons_mvps/init.lua]: ?

Comment (DS-Minetest) [reply mesecons_mvps/init.lua]: ? This works well enough imo, I can't find a simpler way.

**Comment by DS-Minetest** [comment]
Comments addressed.

File: mesecons_mvps/init.lua [hunk 6]
```diff
@@ -205,37 +205,55 @@ end)
...
-			local np = vector.add(obj:getpos(), dir)
-
-			--move only if destination is not solid
-			local nn = minetest.get_node(np)
-			if not ((not minetest.registered_nodes[nn.name])
-			or minetest.registered_nodes[nn.name].walkable) then
-				obj:setpos(np)
+		if ok then
+			local ent = obj:get_luaentity()
+			if not (ent and mesecon.is_mvps_unmov(ent.name)) then
```
Comment (numberZero) [inline mesecons_mvps/init.lua]: Does this work at all?
The correct way seems to be `if ent and not mesecon.is_mvps_unmov(ent.name) then`

File: mesecons_mvps/init.lua (hunk 3 above)
Comment (numberZero) [reply mesecons_mvps/init.lua]: This exact way, but before `for id, obj in pairs(minetest.object_refs) do` and not in the inner loop.

File: mesecons_mvps/init.lua [hunk 7]
```diff
@@ -240,13 +240,18 @@ function mesecon.mvps_move_objects(pos, dir, nodestack)
...
 		if ok then
 			local ent = obj:get_luaentity()
-			if not ent or not mesecon.is_mvps_unmov(ent.name) then
+			if not (ent and mesecon.is_mvps_unmov(ent.name)) then
 				local np = vector.add(obj_pos, dir)
 				-- Move only if destination is not solid or object is inside stack:
 				local nn = minetest.get_node(np)
 				local node_def = minetest.registered_nodes[nn.name]
+				local nodestack_front = (#nodestack - 0.5 + math.abs(pos[dir_k]))
+				nodestack_front = nodestack_front * pos[dir_k]/math.abs(pos[dir_k])
```
Comment (numberZero) [inline mesecons_mvps/init.lua]: Weird math. Can you write something simpler? Remember that you have `dir_l`.
Or let me do that (a bit later).

File: mesecons_mvps/init.lua (hunk 6 above)
Comment (DS-Minetest) [reply mesecons_mvps/init.lua]: This is correct, don't forget players.

Comment (numberZero) [reply mesecons_mvps/init.lua]: Oh, sorry.
Then the previous variant was better. But such clever code needs a comment, at least.
And, well, there is `is_player`.

Comment (DS-Minetest) [reply mesecons_mvps/init.lua]: So, `if obj:is_player() or not mesecon.is_mvps_unmov(obj:get_luaentity().name) then`?

**Pull request merged**